src/timepicker/timepicker.js

```diff
@@ -11,117 +11,47 @@ import * as React from 'react';
 const NOON = DAY / 2;
 
-function dateToSeconds(date: Date) {
-  const seconds = date.getSeconds();
-  const minutes = date.getMinutes() * MINUTE;
-  const hours = date.getHours() * HOUR;
-  return seconds + minutes + hours;
-}
-
-export function secondsToHourMinute(seconds: number) {
-  const d = new Date(seconds * 1000);
-  return [d.getUTCHours(), d.getUTCMinutes()];
-}
-
-function secondsToLabel(seconds, format) {
-  let [hours, minutes] = secondsToHourMinute(seconds);
-  const zeroPrefix = n => (n < 10 ? `0${n}` : n);
-
-  if (format === '12') {
-    const isAfterNoon = seconds >= NOON;
-    if (isAfterNoon) {
-      hours -= 12;
-    }
-
-    if (hours === 0) {
-      hours = 12;
-    }
-
-    return `${hours}:${zeroPrefix(minutes)} ${isAfterNoon ? 'PM' : 'AM'}`;
-  }
-
-  return `${zeroPrefix(hours)}:${zeroPrefix(minutes)}`;
-}
-
-/**
- * Converts a time string, e.g. 10:00, to one or more possible TimePicker
- * options representing that time.
- */
-function stringToOptions(
-  str: string,
-  format: '12' | '24' = '12',
-): Array<OptionT> {
-  // leading zero is optional, AM/PM is optional
-  const twelveHourRegex = /^(1[0-2]|0?[1-9]):([0-5][0-9]) ?([AaPp][Mm]?)?$/;
-  // leading zero is optional
-  const twentyFourHourRegex = /^([0-9]|0[0-9]|1[0-9]|2[0-3]):([0-5][0-9])$/;
-  const regex = format === '12' ? twelveHourRegex : twentyFourHourRegex;
-  const match = str.match(regex);
-  if (!match) {
-    return [];
-  }
-  const hours = Number(match[1]);
-  const minutes = Number(match[2]);
-
-  let hoursMinutes = [];
-  switch (format) {
-    case '24': {
-      hoursMinutes = [{hours, minutes}];
-      break;
-    }
-    case '12':
-    default: {
-      const twelveHours = hours % 12;
-      const meridiem = match[3];
-      // if there's no AM/PM, add both AM and PM options
-      if (!meridiem) {
-        hoursMinutes = [
-          {hours: twelveHours, minutes},
-          {hours: twelveHours + 12, minutes},
-        ];
-      } else {
-        const twentyFourHours =
-          meridiem.toLowerCase()[0] === 'a' ? twelveHours : twelveHours + 12;
-
-        hoursMinutes = [{hours: twentyFourHours, minutes}];
-      }
-      break;
-    }
-  }
-
-  return hoursMinutes.map(({hours, minutes}) => {
-    const secs = hours * 3600 + minutes * 60;
-    return {id: secs, label: secondsToLabel(secs, format)};
-  });
-}
-
-class TimePicker extends React.Component<TimePickerPropsT, TimePickerStateT> {
-  static defaultProps = {
+class TimePicker<T = Date> extends React.Component<
+  TimePickerPropsT<T>,
+  TimePickerStateT,
+> {
+  static defaultProps: TimePickerDefaultPropsT = {
     format: '12',
     step: 900,
     creatable: false,
+    adapter: dateFnsAdapter,
   };
+  dateHelpers: DateHelpers<T>;
 
   state = {steps: [], value: null};
 
   componentDidMount() {
     const steps = this.buildSteps();
-    if (this.props.value instanceof Date) {
+    const {adapter} = this.props;
+    this.dateHelpers = new DateHelpers(adapter);
```

Comment:
The downside of assigning this to a class property is that it will not update in response to props changes. While that is likely a contrived example that should not come up a simple trade-off to an adapter-getter method resolves the issue.

```js
Class Component {
  getDateIOAdapter = () => {
    return this.props.adapter || dateFnsAdapter;
  }
}
```

Reply:
Based on what we discussed on slack, I initially set `DateHelpers` in the constructor, and then reset it in `shouldComponentUpdate`